File: src/main/java/cc/hyperium/cosmetics/butt/ButtCosmetic.java
/*
 *       Copyright (C) 2018-present Hyperium <https://hyperium.cc/>
 *
 *       This program is free software: you can redistribute it and/or modify
 *       it under the terms of the GNU Lesser General Public License as published
 *       by the Free Software Foundation, either version 3 of the License, or
 *       (at your option) any later version.
 *
 *       This program is distributed in the hope that it will be useful,
 *       but WITHOUT ANY WARRANTY; without even the implied warranty of
 *       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *       GNU Lesser General Public License for more details.
 *
 *       You should have received a copy of the GNU Lesser General Public License
 *       along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cc.hyperium.cosmetics.butt;

import cc.hyperium.cosmetics.AbstractCosmetic;
import cc.hyperium.event.InvokeEvent;
import cc.hyperium.event.model.PreCopyPlayerModelAnglesEvent;
import cc.hyperium.event.Priority;
import cc.hyperium.utils.model.IModelPlayer;
import cc.hyperium.purchases.EnumPurchaseType;
import cc.hyperium.purchases.HyperiumPurchase;
import cc.hyperium.purchases.PurchaseApi;

public class ButtCosmetic extends AbstractCosmetic {
    public ButtCosmetic() {
        super(false, EnumPurchaseType.BUTT);
    }

    @InvokeEvent(priority = Priority.HIGH)
    public void preCopy(PreCopyPlayerModelAnglesEvent event) {
        if (!(event.getModel() instanceof IModelPlayer)) return;

        if (!isPurchasedBy(event.getEntity().getUniqueID())) {
            ((IModelPlayer) event.getModel()).getButt().showModel = false;
            return;
        }

        HyperiumPurchase packageIfReady = PurchaseApi.getInstance().getPackageIfReady(event.getEntity().getUniqueID());
        if (packageIfReady == null) return;

        ((IModelPlayer) event.getModel()).getButt().showModel = !packageIfReady.getCachedSettings().isButtDisabled();
        ((IModelPlayer) event.getModel()).getButt().offsetY = event.getEntity().isSneaking() ? -.45F : -.4F;
        ((IModelPlayer) event.getModel()).getButt().offsetZ = event.getEntity().isSneaking() ? .3F : 0F;
    }
}

File: src/main/kotlin/cc/hyperium/cosmetics/butt/ButtCosmetic.kt
/*
 *       Copyright (C) 2018-present Hyperium <https://hyperium.cc/>
 *
 *       This program is free software: you can redistribute it and/or modify
 *       it under the terms of the GNU Lesser General Public License as published
 *       by the Free Software Foundation, either version 3 of the License, or
 *       (at your option) any later version.
 *
 *       This program is distributed in the hope that it will be useful,
 *       but WITHOUT ANY WARRANTY; without even the implied warranty of
 *       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *       GNU Lesser General Public License for more details.
 *
 *       You should have received a copy of the GNU Lesser General Public License
 *       along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package cc.hyperium.cosmetics.butt

import cc.hyperium.cosmetics.AbstractCosmetic
import cc.hyperium.event.InvokeEvent
import cc.hyperium.event.Priority
import cc.hyperium.event.model.PreCopyPlayerModelAnglesEvent
import cc.hyperium.purchases.EnumPurchaseType
import cc.hyperium.purchases.PurchaseApi
import cc.hyperium.utils.model.IModelPlayer

class ButtCosmetic : AbstractCosmetic(false, EnumPurchaseType.BUTT) {
    @InvokeEvent(priority = Priority.HIGH)
    fun preCopy(event: PreCopyPlayerModelAnglesEvent) {
        if (event.model !is IModelPlayer) return
        if (!isPurchasedBy(event.entity.uniqueID)) {
            (event.model as IModelPlayer).butt.showModel = false
            return
        }

        val packageIfReady = PurchaseApi.getInstance().getPackageIfReady(event.entity.uniqueID) ?: return
        (event.model as IModelPlayer).butt.showModel = !packageIfReady.cachedSettings.isButtDisabled
        (event.model as IModelPlayer).butt.offsetY = if (event.entity.isSneaking) -.45f else -.4f
        (event.model as IModelPlayer).butt.offsetZ = if (event.entity.isSneaking) .3f else 0f
    }
}
